lib/util/stream_utils.js

```diff
@@ -30,55 +30,117 @@ shaka.util.StreamUtils = class {
-      }
+    // Now organize variants into buckets by codecs.
+    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+    let variantsByCodecs = StreamUtils.getVariantsByCodecs(variants);
+    variantsByCodecs = StreamUtils.filterVariantsByHeight(variantsByCodecs);
+
+    const bestCodecs = StreamUtils.findBestCodecs(variantsByCodecs);
 
-      let baseAudioCodec = '';
-      if (variant.audio) {
-        baseAudioCodec = MimeUtils.getCodecBase(variant.audio.codecs);
+    // Filter out any variants that don't match, forcing AbrManager to choose
+    // from the most efficient variants possible.
+    manifest.variants = manifest.variants.filter((variant) => {
+      const codecs = StreamUtils.getGroupVariantCodecs(variant);
+      if (codecs == bestCodecs) {
+        return true;
       }
 
-      return baseVideoCodec + '-' + baseAudioCodec;
-    }
+      shaka.log.debug('Dropping Variant (better codec available)', variant);
+      return false;
+    });
+  }
 
-    // Now organize variants into buckets by codecs.
-    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+  /**
+  * Get variants by codecs.
+  *
+  * @param {!Array<shaka.extern.Variant>} variants
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  */
+  static getVariantsByCodecs(variants) {
     const variantsByCodecs = new shaka.util.MultiMap();
     for (const variant of variants) {
-      const group = variantCodecs(variant);
+      const group = shaka.util.StreamUtils.getGroupVariantCodecs(variant);
       variantsByCodecs.push(group, variant);
     }
 
-    // Compute the average bandwidth for each group of variants.
-    // Choose the lowest-bandwidth codecs.
-    let bestCodecs = null;
+    return variantsByCodecs;
+  }
+
+  /**
+  * Filters variants with missing heights by all codecs.
+  *
+  * Average bandwidth of H.265 codecs (vp9) is higher than H.264 codecs (avc1).
+  * H.265 has 4K support and more variants.
+  *
+  * @param {!shaka.util.MultiMap.<shaka.extern.Variant>} variantsByCodecs
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  */
+  static filterVariantsByHeight(variantsByCodecs) {
+    const heightByCodecs = new Map();
+    variantsByCodecs.forEach((codecs, variants) => {
+      for (const variant of variants) {
+        if (!variant.video) {
+          continue;
+        }
+
+        const key = variant.video.height;
+        if (!heightByCodecs.has(key)) {
+          heightByCodecs.set(key, new Set());
+        }
+
+        heightByCodecs.get(key).add(codecs);
+      }
+    });
+
+    const countCodecs = variantsByCodecs.size();
+
+    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+    const filteredVariantsByCodecs = new shaka.util.MultiMap();
+    variantsByCodecs.forEach((codecs, variants) => {
+      for (const variant of variants) {
+        if (
+          variant.video &&
+          heightByCodecs.has(variant.video.height) &&
+          heightByCodecs.get(variant.video.height).size !== countCodecs
```

Comment:
So you eliminate a variant if its height is not available in all codecs?  That sounds like the opposite of what is needed to solve this problem.  Am I misreading the code?

If you have this, as you described in the body of the PR:

avc1: 240, 360, 480, 720, 1080
vp9: 240, 360, 480, 720, 1080, 1440, 2160

`countCodecs` will be 2, but `heightByCodecs.get(1440)` will be 1, so you skip it.  This will result in the high res codecs being eliminated.  This will "correct" the average bandwidth of vp9 in your example to be lower than h264, but only by forcing 2K and 4K to be skipped.

It seems like the better strategy is to ask what is the maximum resolution for each supported codec, then choose the codec with the highest maximum resolution.  If there is a tie, then fall back to choosing the lowest average bandwidth.

Reply:
Actually, I think I made a mistake in reading through this.  This only eliminates variants in the structures used to choose codecs, before the average bandwidth is calculated.  This doesn't eliminate them from the manifest structure, so they can still be played.  My mistake.

That said, I think my suggestion might still be a clearer way to look at the problem: Choose highest res, then fall back to lowest bandwidth if there's a tie.  You could track the max height overall and the max height per codec, instead of a map of heights to codecs, then remove the codecs whose max height isn't equal to the max height overall.

Does that make sense?  I think the outcome would be the same, but I think it will be easier to understand that way.

Reply:
Yes, we only compare the same resolutions:
avc1: 240, 360, 480, 720, 1080
vp9: 240, 360, 480, 720, 1080, ~~1440, 2160~~

`It seems like the better strategy is to ask what is the maximum resolution for each supported codec, then choose the codec with the highest maximum resolution. If there is a tie, then fall back to choosing the lowest average bandwidth.`

I want to make a simple algorithm. In your version, you will need to find the maximum of the same resolution. But you need to take into account the additional frameRate and possibly something else. The bandwidth of 1080p@30 is less than 1080p@60.

Reply:
I don't see why you would consider my version more complex.  Could you elaborate?

And wouldn't you need to deal with framerates in your version, too?  For example:

h264: 480, 720, 1080p30
vp9: 480, 720, 1080p30, 1080p60

If you only look at height, the 1080p60 variant would be included in the average for vp9 with these inputs, and the average would skew high on vp9.  Depending on the numbers, it may turn out be higher than the h264 version with these inputs, leading to the same problem we started with.

I think you make a good point about framerate being a consideration, but I think framerate should be a factor, no matter whether we use your algorithm or mine.

Where I used max height in my example, you could replace height with a combination of height and framerate, such as `(height * 1000) + framerate`.  The same substitution could be made in yours, too, for the purposes of filtering.

What do you think?